apollo-router-core/src/query_planner/mod.rs

```diff
@@ -253,66 +270,129 @@ impl PlanNode {
+        requires: Option<Vec<Selection>>,
+
+        /// The variables that are used for the subgraph fetch.
+        variable_usages: Vec<String>,
+
+        /// The GraphQL subquery that is used for the fetch.
+        operation: String,
+    }
+
+    struct Variables {
+        variables: Map<String, Value>,
+        paths: Vec<Path>,
+    }
+
+    impl Variables {
+        fn new<'a>(
+            requires: Option<&Vec<Selection>>,
+            variable_usages: &[String],
+            data: &Value,
+            current_dir: &'a Path,
+            request: &Arc<Request>,
+            schema: &Arc<Schema>,
+        ) -> Result<Variables, FetchError> {
+            if let Some(requires) = requires {
+                let mut variables = Object::with_capacity(1 + variable_usages.len());
+                variables.extend(variable_usages.iter().filter_map(|key| {
+                    request.variables.as_ref().map(|v| {
+                        v.get(key)
+                            .map(|value| (key.clone(), value.clone()))
+                            .unwrap_or_default()
+                    })
+                }));
+
+                let mut paths = Vec::new();
+                let mut values = Vec::new();
+                data.select_values_and_paths(current_dir, |_path, value| {
+                    paths.push(_path);
+                    values.push(value)
+                })?;
+
+                let representations = Value::Array(
+                    values
+                        .into_iter()
+                        .flat_map(|value| match value {
+                            Value::Object(content) => {
+                                select_object(content, requires, schema).transpose()
+                            }
+                            _ => Some(Err(FetchError::ExecutionInvalidContent {
+                                reason: "not an object".to_string(),
+                            })),
+                        })
+                        .collect::<Result<Vec<_>, _>>()?,
                 );
-                let representations = selection::select(&response, current_dir, requires, &schema)?;
                 variables.insert("representations".into(), representations);
+
+                Ok(Variables { variables, paths })
+            } else {
+                Ok(Variables {
+                    variables: variable_usages
+                        .iter()
+                        .filter_map(|key| {
+                            request
+                                .variables
+                                .as_ref()
+                                .map(|v| v.get(key).map(|value| (key.clone(), value.clone())))
+                                .unwrap_or_default()
+                        })
+                        .collect::<Object>(),
+                    paths: Vec::new(),
+                })
             }
+        }
+    }
+
+    impl FetchNode {
+        pub(crate) async fn fetch_node<'a>(
```

Comment:
:yarn: This lifetime looks unnecessary.

I think it's only used to make self life as long as current_dir. All the rest can use anonymous lifetime.

Reply:
Any attempt at making one of the lifetimes anonymous results in the following error:

```
   Compiling apollo-router-core v0.1.0-alpha.2 (/home/geal/dev/router/apollo-router-core)
error[E0700]: hidden type for `impl Trait` captures lifetime that does not appear in bounds
   --> apollo-router-core/src/query_planner/mod.rs:374:14
    |
374 |         ) -> Result<Value, FetchError> {
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: hidden type `impl futures::Future` captures lifetime smaller than the function body
   --> apollo-router-core/src/query_planner/mod.rs:374:14
    |
374 |         ) -> Result<Value, FetchError> {
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0700`.
```

Reply:
lifetimes are hard xD